src/Biogeochemistry.jl

```diff
@@ -0,0 +1,181 @@
+import Oceananigans.Forcings: regularize_forcing, maybe_constant_field
+
+#####
+##### Generic fallbacks for biogeochemistry
+#####
+
+"""
+Update tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_tendencies!(bgc, model) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_biogeochemical_state!(bgc, model) = nothing
+
+@inline biogeochemical_drift_velocity(bgc, val_tracer_name) = nothing
+@inline biogeochemical_advection_scheme(bgc, val_tracer_name) = nothing
+
+#####
+##### Default (discrete form) biogeochemical source
+#####
+
+abstract type AbstractBiogeochemistry end
+
+@inline function biogeochemistry_rhs(i, j, k, grid, bgc, val_tracer_name::Val{tracer_name}, clock, fields) where tracer_name
+    U_drift = biogeochemical_drift_velocity(bgc, val_tracer_name)
+    scheme = biogeochemical_advection_scheme(bgc, val_tracer_name)
+    src = biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields)
+    c = fields[tracer_name]
+        
+    return src - div_Uc(i, j, k, grid, scheme, U_drift, c)
+end
+
+@inline biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields) =
+    bgc(i, j, k, grid, val_tracer_name, clock, fields)
+
+@inline (bgc::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+#####
+##### Continuous form biogeochemical source
+#####
+ 
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractContinuousFormBiogeochemistry <: AbstractBiogeochemistry end
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{1}) =
+    @inbounds (fields[names[1]][i, j, k],)
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{2}) =
+    @inbounds (fields[names[1]][i, j, k],
+               fields[names[2]][i, j, k])
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{N}) where N =
+    @inbounds ntuple(n -> fields[names[n]][i, j, k], Val(N))
+
+@inline function biogeochemical_transition(i, j, k, grid, bgc::AbstractContinuousFormBiogeochemistry,
+                                           val_tracer_name, clock, fields)
+
+    names_to_extract = tuple(required_biogeochemical_tracers(bgc)...,
+                             required_biogeochemical_auxiliary_fields(bgc)...)
+
+    fields_ijk = extract_biogeochemical_fields(i, j, k, grid, fields, names_to_extract)
+
+    x = xnode(Center(), Center(), Center(), i, j, k, grid)
+    y = ynode(Center(), Center(), Center(), i, j, k, grid)
+    z = znode(Center(), Center(), Center(), i, j, k, grid)
+
+    return bgc(val_tracer_name, x, y, z, clock.time, fields_ijk...)
+end
+
+@inline (bgc::AbstractContinuousFormBiogeochemistry)(val_tracer_name, x, y, z, t, fields...) = zero(x)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
+
+tracernames(tracers) = keys(tracers)
+tracernames(tracers::Tuple) = tracers
```

Comment:
Is this function defined in Utils?

Reply:
Just checked and its in `field_tuples`

Reply:
I guess its simple enough to keep here...